core/Symbols.cc

```diff
@@ -1062,7 +1062,7 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
     ENFORCE(lastClassType != nullptr, "Last element of sealedSubclasses must be ClassType");
     auto subclass = lastClassType->symbol.data(gs)->attachedClass(gs);
     ENFORCE(subclass.exists());
-    result = Types::any(gs, make_type<ClassType>(subclass), result);
+    result = Types::any(gs, subclass.data(gs)->externalType(gs), result);
```

Comment:
From dmitry: this approach of just calling `externalType` is just wrong. This function shouldn't be operating on symbol only symbols.

Instead, it should receive the `from` type (like `List[Integer]`) and use a constraint solver to solve for the type args that the sealed subclasses will need to be applied to.

He said something about our current constraint solver being tuned for inference, which means sometimes it gives solutions that wouldn't be good enough for this use case.